pkg/autoscaler/autoscaler.go

```diff
@@ -732,6 +518,96 @@ var _ = Describe("[Feature:Machines] Autoscaler should", func() {
 			}, framework.WaitMedium, pollingInterval).Should(BeTrue())
 		})
 
+		It("scales up and down while respecting MaxNodesTotal [Slow][Serial]", func() {
+			By("Creating 1 MachineSet with 1 replica")
+			var transientMachineSet *mapiv1beta1.MachineSet
+			targetedNodeLabel := fmt.Sprintf("%v-scale-updown", autoscalerWorkerNodeRoleLabel)
+			machineSetParams := framework.BuildMachineSetParams(client, 1)
+			machineSetParams.Labels[targetedNodeLabel] = ""
+			transientMachineSet, err = framework.CreateMachineSet(client, machineSetParams)
+			Expect(err).ToNot(HaveOccurred())
+			cleanupObjects[transientMachineSet.GetName()] = transientMachineSet
+
+			By("Waiting for all Machines in the MachineSet to enter Running phase")
+			framework.WaitForMachineSet(client, transientMachineSet.GetName())
+
+			// To exercise the MaxNodesTotal mechanism we want to craft this
+			// MachineAutoscaler so that its maximum size would cause the cluster
+			// to grow beyond the MaxNodesTotal. This is calculated as the
+			// MaxNodesTotal - current number of nodes
+			// + 1 (for the initial size of the MachineSet)
+			// + 1 (to exceed the maximum)
+			By("Calculating maximum size for MachineAutoscaler")
+			nodes, err := framework.GetNodes(client)
```

Comment:
What happens if some of these nodes disappear shortly after you list them, means the later numbers won't line up right? (eg previous test has nodes going away?)

Reply:
it would definitely get messed up, i marked this test as "[serial]". i think when we are testing if the autoscaler will stop at the max cluster size we need to run that test by itself.

this kinda ties into something we had talked about before, namely the tests cleaning up after themselves. currently, we shouldn't have an issue with this test breaking in the middle like this, but it could. i can see that switching to checking the cluster size will probably make this interaction less harmful, especially if i just make the machineset be able to grow as large as needed.